cborg/tests/Tests/FlatTerm.hs

```diff
@@ -2,120 +2,113 @@ module Tests.FlatTerm
-prop_validateValidFlat (ValidFlat ts) = property $ validFlatTerm ts
-
-prop_validateInValidFlat :: InValidFlat -> Property
-prop_validateInValidFlat (InValidFlat ts) = property $ validFlatTerm ts == False
+instance Arbitrary TermToken where
+  arbitrary = oneof
+    [ TkInt <$> arbitrary
+    , TkInteger <$> arbitrary
+    , TkBytes . BS.pack <$> arbitrary 
+    , pure TkBytesBegin
+    , TkString . T.pack <$> arbitrary
+    , pure TkStringBegin
+    , TkListLen <$> arbitrary
+    , pure TkListBegin
+    , TkMapLen <$> arbitrary
+    , pure TkMapBegin
+    , pure TkBreak
+    , TkTag <$> arbitrary
+    , TkBool <$> arbitrary
+    , pure TkNull
+    , TkSimple <$> arbitrary
+    , TkFloat16 <$> arbitrary
+    , TkFloat32 <$> arbitrary
+    , TkFloat64 <$> arbitrary ]
+
+-- | Converts a FlatTerm to Encoding
+tokenToEncoding :: TermToken -> Encoding
+tokenToEncoding (TkInt n)      = encodeInt n  
+tokenToEncoding (TkInteger i)  = encodeInteger i
+tokenToEncoding (TkBytes bs)   = encodeBytes bs
+tokenToEncoding TkBytesBegin   = encodeBytesIndef
+tokenToEncoding (TkString txt) = encodeString txt
+tokenToEncoding TkStringBegin  = encodeStringIndef
+tokenToEncoding (TkListLen w)  = encodeListLen w
+tokenToEncoding TkListBegin    = encodeListLenIndef
+tokenToEncoding (TkMapLen w)   = encodeMapLen w
+tokenToEncoding TkMapBegin     = encodeMapLenIndef 
+tokenToEncoding TkBreak        = encodeBreak 
+tokenToEncoding (TkTag w64)    = encodeTag64 w64
+tokenToEncoding (TkBool bool)  = encodeBool bool
+tokenToEncoding TkNull         = encodeNull
+tokenToEncoding (TkSimple w8)  = encodeSimple w8
+tokenToEncoding (TkFloat16 f)  = encodeFloat16 f
+tokenToEncoding (TkFloat32 f)  = encodeFloat f
+tokenToEncoding (TkFloat64 d)  = encodeDouble d
+
+-- | Changes a FlatTerm to List of Word8.    
+termTokensToWord :: [TermToken] -> [Word8]
+termTokensToWord ts = LBS.unpack bytes
+  where
+    bytes = toLazyByteString enc
+    enc = mconcat $ tokenToEncoding <$> ts 
+
+-- | Given a FlatTerm returns a tuple Decoded by Reference and Implementation
+-- decoders respectively as First and Second element of the tuple.
+refImpDecode :: [TermToken] -> (Maybe Imp.Term, Maybe Imp.Term)
+refImpDecode ts = (((Imp.canonicaliseTerm . toRefTerm ) <$> maybeRdecoded), (Imp.canonicaliseTerm <$> maybeIdecoded ))
+  where
+    impDecoded = Imp.runDecoder Imp.decodeTerm (termTokensToWord ts)
+    refDecoded = fromFlatTerm decodeTerm ts
+    maybeRdecoded = either (const Nothing) Just refDecoded 
+    maybeIdecoded = case impDecoded of
+      Just (term, []) -> Just term
+      _               -> Nothing
+
+-- | Given a FlatTerm checks whether the results of Refernce and Implementation
+-- decoders tally or not, also matches the result with @validFlatTerm@ function.
+prop_validFlatTerm :: [TermToken] -> Property
+prop_validFlatTerm ts = property $ case refImpDecode ts of
+  (Just x, Just y)   -> x == y && validFlatTerm ts
+  (Nothing, Nothing) -> True && (not $ validFlatTerm ts)
```

Comment:
In that position `True` does absolutely nothing. You should drop it an just leave the `not (validFlatTerm ts)`

Reply:
You have the logic right here (ie `prop_validFlatTerm` and `refImpDecode`) but the structure is a bit weird. Specifically, you have an function that encodes a pair two different ways and returns them as a tuple, when a large chunk of that could just be inlined in the property. Once you inline like that, you probably don't need `maybeRdecoded` you can just pattern match on an `(Maybe a, Either e a)` pair.

Reply:
While i think the above is logically correct, I think that the QC generated `[TermToken] ` data will be invalid for all but the most trivial cases. It would be much nicer if we could manipulate this so that it mostly generated terms that were valid say 80% of the time and invalid 20% if the time. This would give us much better assurance that this property is getting thoroughly tested.

Reply:
Implemented the requested changes!